### PR TITLE
Resource Loading Performance Updates

### PR DESCRIPTION
**Pull request by codewithkyle**
### Added

-   web component `connection-type` attribute support -- defaults to `2g`

### Fixed

-   CSS & JS loading performance [#77](https://github.com/Pageworks/djinnjs/issues/77)

### CODE REVIEW

**Comment by codewithkyle** [comment]
@bamccaul-dev The important thing to review in this PR is the new `connection-type` attribute. By default it is set to `2g` meaning anyone browsing a website on a `slow-2g` connection will not download any JS for the default web components. If a web component is critical a `connection-type="slow-2g"` attribute must be applied to the custom element.

I chose to set the default minimum as a `2g` connection since I believe we both agree that JavaScript should be considered icing on the cake instead of a primary ingredient. For users on the slowest possible connection we should deliver a hyper-lightweight version of the site.

One additional improvement I should see being useful would be a snackbar notification informing the user on a `slow-2g` connection that they are viewing a lightweight version of the site and some features may be unavailable with the suggestion of connecting to WIFI and reloading for the full experience.

**Comment by codewithkyle** [comment]
Something else to consider is how do we handle dead components on slow connections? If a user is on a `slow-2g` connection what happens to the web components that will never load? Should there be a `purgeable` attribute that tells DjinnJS to call `customElement.remove()` removing the HTML from the DOM?

The use case for this is limited but could be useful in certain situations where we are utilizing client-side rendering.

**Comment by bamccaul-dev** [comment]
@codewithkyle A thought I had with the 2g snack bar: What if we let the user decide if they want to wait for the "regular version" of the site? Some people might be willing to let the site load for however long it is. I can see that being wanted.

**Comment by codewithkyle** [comment]
It would be `slow-2g` is 114 Kbps (0.1 Mbps) not 2g connection.

**Comment by codewithkyle** [comment]
I could see merging the purgeable idea with the snackbar. Inform the user they are viewing a lightweight version of the website and the user can choose to continue with the lightweight or wait for the full version. If they choose the lightweight version we can purge any web components with `purgeable` but if they choose to continue we will drop the connection requirements.

**Comment by codewithkyle** [comment]
We could also bump it up to a `3g` default and prompt users with the full-version snackbar on a `2g` or `slow-2g` connection.

**Comment by bamccaul-dev** [comment]
I like that better. I think that would be more ideal as a default.

**Comment by codewithkyle** [comment]
@bamccaul-dev The new `removable` attribute has been added and the `connection-type` attribute has been reworked as the `required-connection` attribute. I've also implemented the lightweight notification functionality we discussed in our meeting.

File: src/core/runtime.ts [hunk 1]
```diff
@@ -131,6 +136,72 @@ class Runtime {
...
+                return;
+            } else if (sessionChoice === "2") {
+                this.removePurgeableComponents();
+                resolve();
+                return;
+            }
+
+            if (env.connection === "2g" || env.connection === "slow-2g" || env.connection === "3g") {
+                notify({
+                    message: "You are viewing the lightweight verison of this website. Would you like to load the full experience instead?",
```
Comment (bamccaul-dev) [inline src/core/runtime.ts:154]: Perhaps this should go into a strings config file for easier translation?

**Review by bamccaul-dev — APPROVED** [review]

**Pull request merged**